limit_defaults.go

```diff
@@ -1,166 +1,425 @@
+// If memory is smaller than 128 MB, the base configuration will be used.
+//
+func (cfg *ScalingLimitConfig) Scale(memory int64, numFD int) LimitConfig {
+	var scaleFactor int
+	if memory > 128<<20 {
+		scaleFactor = int((memory - 128<<20) >> 20)
+	}
+	lc := LimitConfig{
+		System:              scale(cfg.SystemBaseLimit, cfg.SystemLimitIncrease, scaleFactor, numFD),
+		Transient:           scale(cfg.TransientBaseLimit, cfg.TransientLimitIncrease, scaleFactor, numFD),
+		ServiceDefault:      scale(cfg.ServiceBaseLimit, cfg.ServiceLimitIncrease, scaleFactor, numFD),
+		ServicePeerDefault:  scale(cfg.ServicePeerBaseLimit, cfg.ServicePeerLimitIncrease, scaleFactor, numFD),
+		ProtocolDefault:     scale(cfg.ProtocolBaseLimit, cfg.ProtocolLimitIncrease, scaleFactor, numFD),
+		ProtocolPeerDefault: scale(cfg.ProtocolPeerBaseLimit, cfg.ProtocolPeerLimitIncrease, scaleFactor, numFD),
+		PeerDefault:         scale(cfg.PeerBaseLimit, cfg.PeerLimitIncrease, scaleFactor, numFD),
+		Conn:                scale(cfg.ConnBaseLimit, cfg.ConnLimitIncrease, scaleFactor, numFD),
+		Stream:              scale(cfg.StreamBaseLimit, cfg.ConnLimitIncrease, scaleFactor, numFD),
+	}
+	if cfg.ServiceLimits != nil {
+		lc.Service = make(map[string]BaseLimit)
+		for svc, l := range cfg.ServiceLimits {
+			lc.Service[svc] = scale(l.BaseLimit, l.BaseLimitIncrease, scaleFactor, numFD)
+		}
+	}
+	if cfg.ProtocolLimits != nil {
+		lc.Protocol = make(map[protocol.ID]BaseLimit)
+		for proto, l := range cfg.ProtocolLimits {
+			lc.Protocol[proto] = scale(l.BaseLimit, l.BaseLimitIncrease, scaleFactor, numFD)
+		}
+	}
+	if cfg.PeerLimits != nil {
+		lc.Peer = make(map[peer.ID]BaseLimit)
+		for p, l := range cfg.PeerLimits {
+			lc.Peer[p] = scale(l.BaseLimit, l.BaseLimitIncrease, scaleFactor, numFD)
+		}
+	}
+	if cfg.ServicePeerLimits != nil {
+		lc.ServicePeer = make(map[string]BaseLimit)
+		for svc, l := range cfg.ServicePeerLimits {
+			lc.ServicePeer[svc] = scale(l.BaseLimit, l.BaseLimitIncrease, scaleFactor, numFD)
+		}
+	}
+	if cfg.ProtocolPeerLimits != nil {
+		lc.ProtocolPeer = make(map[protocol.ID]BaseLimit)
+		for p, l := range cfg.ProtocolPeerLimits {
+			lc.ProtocolPeer[p] = scale(l.BaseLimit, l.BaseLimitIncrease, scaleFactor, numFD)
+		}
+	}
+	return lc
 }
 
-func (cfg *DefaultLimitConfig) WithSystemMemory(memFraction float64, minMemory, maxMemory int64) DefaultLimitConfig {
-	refactor := memFraction / cfg.SystemMemory.MemoryFraction
-	r := *cfg
-	r.SystemMemory.MemoryFraction = memFraction
-	r.SystemMemory.MinMemory = minMemory
-	r.SystemMemory.MaxMemory = maxMemory
-	r.TransientMemory.MemoryFraction *= refactor
-	r.ServiceMemory.MemoryFraction *= refactor
-	r.ServicePeerMemory.MemoryFraction *= refactor
-	r.ProtocolMemory.MemoryFraction *= refactor
-	r.ProtocolPeerMemory.MemoryFraction *= refactor
-	r.PeerMemory.MemoryFraction *= refactor
-	return r
+func (cfg *ScalingLimitConfig) AutoScale() LimitConfig {
+	return cfg.Scale(
+		int64(memory.TotalMemory())/8,
```

Comment:
why 8? I saw 1/8th mention above, I just don't follow where this comes from.

Reply:
You don't want libp2p to consume all the memory that you have. The 1/8 is totally arbitrary, but intended as a sane default. `AutoScale` is only intended as a convenience function, if you want more control, `Scale` is what you'd use.